### PR TITLE
oshmem: Reduce the service memory footprint

### PR DESCRIPTION
**Pull request by sb17v**
Current implementation is maintaining a static cache of 32 entries
per remote peer containing memory key information.
In practice, only 2-3 entries are used. At scale, however this leads
to a significant overhead.

Example:
* one cache entry is 40B
* with default 32 entries per peer it is 1280B per peer.
* on cluster with 1000 nodes and 40 PPN it results in 51MB per process
* Cumulatively, per node (given 40PPN) its 51MB x 40 = 2GB of overhead

This commit introduces dynamic cache that contains only required number
of entries. With 2 entries by default this leads to 16x reduction in the
cache memory footprint.

Future optimization of the cache entry layout is planned.
Will be PR'd as a follow-up.

Co-authored-by: Artem Y. Polyakov <artemp@nvidia.com>
Signed-off-by: Subhadeep Bhattacharya <subhadeepb@nvidia.com>
Signed-off-by: Artem Polyakov <artemp@nvidia.com>

### CODE REVIEW

**Comment by sb17v** [comment]
@janjust @karasevb FYI

File: oshmem/mca/spml/ucx/spml_ucx.c [hunk 1]
```diff
@@ -104,6 +104,155 @@ int mca_spml_ucx_enable(bool enable)
...
+        return rc;
+    }
+
+    if (mkey->u.data) {
+        err = ucp_ep_rkey_unpack(ucx_ctx->ucp_peers[pe].ucp_conn,
+                                    mkey->u.data,
+                                    &((*ucx_mkey)->rkey));
+        if (UCS_OK != err) {
+            SPML_UCX_ERROR("failed to unpack rkey: %s", ucs_status_string(err));
+            return rc;
```
Comment (karasevb) [inline oshmem/mca/spml/ucx/spml_ucx.c]: Should return `OSHMEM_ERROR` instead of `rc`?

**Comment by manjugv** [comment]
@sb17v @artemp Please make sure that you run correctness benchmarks - OSU, SOS, and vendor tests (https://github.com/openshmem-org/tests-mellanox).

File: oshmem/mca/spml/ucx/spml_ucx.c (hunk 1 above)
Comment (sb17v) [reply oshmem/mca/spml/ucx/spml_ucx.c]: Yes. Fixed.

**Comment by artpol84** [comment]
Thanks @manjugv
We did SOS, but will do OSU and the vendor one as well.

**Pull request merged**